tests/test_csc.py

```diff
@@ -0,0 +1,302 @@
+        disconnect the device.
+        """
+        self.patch.return_value.AVS_GetScopeData.side_effect = None
+        self.patch.return_value.AVS_GetScopeData.return_value = AvsReturnCode.ERR_INVALID_MEAS_DATA.value
+        async with Harness(initial_state=salobj.State.ENABLED) as harness:
+            # Check that we are properly in ENABLED at the start
+            await self.check_summaryState(harness.remote, salobj.State.ENABLED)
+
+            msg = "Failed to take exposure"
+            with salobj.assertRaisesAckError(ack=salobj.SalRetCode.CMD_FAILED, result_contains=msg):
+                await harness.remote.cmd_expose.set_start(timeout=STD_TIMEOUT, duration=.5)
+            # the exposure state should be Integrating during the exposure
+            await self.check_exposureState(harness.remote, ExposureState.INTEGRATING)
+            await self.check_summaryState(harness.remote, salobj.State.FAULT)
+            error = await harness.remote.evt_errorCode.next(flush=False, timeout=STD_TIMEOUT)
+            errorMsg = str(AvsReturnError(AvsReturnCode.ERR_INVALID_MEAS_DATA.value, "GetScopeData"))
+            self.assertIn(errorMsg, error.errorReport)
+            # Going into FAULT should close the device connection.
+            self.assertIsNone(harness.csc.device)
+            # the exposure state should be FAILED after a failed exposure
+            await self.check_exposureState(harness.remote, ExposureState.FAILED)
+
+    async def test_cancelExposure(self):
+        """Test that we can stop an active exposure, and that the exposureState
+        is changed appropriately.
+        """
+        async with Harness(initial_state=salobj.State.ENABLED) as harness:
+            # Check that we are properly in ENABLED at the start
+            await self.check_summaryState(harness.remote, salobj.State.ENABLED)
+
+            duration = 5  # seconds
+            task = asyncio.create_task(harness.remote.cmd_expose.set_start(timeout=STD_TIMEOUT + duration,
+                                                                           duration=duration))
+            await asyncio.sleep(0.1)  # give the event loop time to start
+            await harness.remote.cmd_cancelExposure.set_start(timeout=STD_TIMEOUT)
+            with salobj.assertRaisesAckError(ack=salobj.SalRetCode.CMD_ABORTED):
+                await task
+            # flush out the Integrating state from during the exposure
+            await self.check_exposureState(harness.remote, ExposureState.INTEGRATING)
+            # the exposure state should be CANCELLED after a cancelled exposure
+            await self.check_exposureState(harness.remote, ExposureState.CANCELLED)
+
+    async def test_simulator(self):
+        """Test that we can turn the simulation mode on and off.
+
+        We know we are in simulation mode if the device "library" loaded by
+        the device is a different mock than `self.patch`.
```

Comment:
I find this confusing. When not in simulation mode why is "is a diiferent mock" relevant? You don't have a mock at all.

Reply:
When not in simulation mode *in the test suite* we have the mock patch defined in the test's `setUp`. How should I clarify that here?

Reply:
I don't know. I don't fully understand your clarification. Mocks make my head hurt.

Reply:
When running the tests, there is a mock patch of the library. Whether or not you're running in simulation mode, when you're running the tests, libavs is a mock patch (it has to be, otherwise you'd have to have the device connected and you wouldn't be able to force certain behavior).

When turning on simulation mode, there is a *different instance* of the mock patch of the library. Whether or not you're running a test, the simulator has its own mock patch.

The test checks that the mock patch of the library held by the `csc.device` is different from the test mock patch.

Reply:
We iterated offline on some updated docstrings, which I've added.